src/storage/GenericVersionedStorage.ts

```
@@ -0,0 +1,188 @@
+import { DeepTokensMap, ThemeObjectsList } from '@/types';
+import { SingleToken } from '@/types/tokens';
+import { RemoteTokenStorage, RemoteTokenStorageFile } from './RemoteTokenStorage';
+import { singleFileSchema } from './schemas/singleFileSchema';
+import { GenericVersionedStorageFlow } from '@/types/StorageType';
+import { tokensMapSchema } from './schemas/tokensMapSchema';
+import { themeObjectSchema } from './schemas/themeObjectSchema';
+
+const genericVersionedSchema = singleFileSchema.extend({
+  values: z.record(tokensMapSchema),
+  version: z.string().optional(),
+  $themes: z.array(themeObjectSchema).optional(),
+  updatedAt: z.number().optional(),
+});
+
+type GenericVersionedMeta = {
+  version: string
+  updatedAt: string
+};
+
+export type GenericVersionedData = GenericVersionedMeta & {
+  values: Record<string, Record<string, SingleToken<false> | DeepTokensMap<false>>>
+  $themes?: ThemeObjectsList
+};
+
+export type GenericVersionedAdditionalHeaders = Array<{
+  name: string,
+  value: string
+}>;
+
+const flattenHeaders = (headers: GenericVersionedAdditionalHeaders): Array<[key: string, val: string]> => headers.map((x) => [x.name, x.value]);
+
+// Life cycle is
+
+// POST - Create storage at endpoint. Storage may or may not already exist
+// Get - Retrieve tokens
+// PUT - Push tokens to endpoint
+
+export class GenericVersionedStorage extends RemoteTokenStorage<GenericVersionedMeta> {
+  private url: string;
+
+  private defaultHeaders: Headers;
+
+  public static async create(url: string, updatedAt: string, flow: GenericVersionedStorageFlow, headers: GenericVersionedAdditionalHeaders = []): Promise<false | {
+    metadata: { id: string }
+  }> {
+    let body: string | undefined = JSON.stringify({
+      version: pjs.plugin_version,
+      updatedAt,
+      values: {
+        options: {},
+      },
+    }, null, 2);
+    let method: string;
+    switch (flow) {
+      case GenericVersionedStorageFlow.READ_ONLY:
```

We should probably just throw an error if we are trying to create in read only mode

---

No that would be a valid case. The "creation" would check against the remote to pull the latest tokens at the time of setup.